himbaechel/uarch/ng-ultra/ng_ultra.cc

```diff
@@ -0,0 +1,1027 @@
+            if (ctx->getBoundBelCell(ctx->getBelByLocation(Loc(loc.x+fix,loc.y,BEL_XFIFO_Z)))) return false;
+            // If we used coresponding FIFO we can not use CDC
+            if (ctx->getBoundBelCell(ctx->getBelByLocation(Loc(loc.x+fix,loc.y,BEL_FIFO_Z + index)))) return false;
+        }
+    }
+    return true;
+}
+
+// Bel bucket functions
+IdString NgUltraImpl::getBelBucketForCellType(IdString cell_type) const
+{
+    if (cell_type.in(id_IOP,id_IP,id_OP,id_IOTP,id_ITP,id_OTP))
+        return ctx->idf("IOP/IOTP");
+    else if (cell_type.in(id_BFR, id_DFR, id_DDFR))
+        return ctx->idf("DFR/DDFR");
+    else if (cell_type.in(id_RF, id_RFSP))
+        return id_RF;
+    else if (cell_type.in(id_XHRF, id_XWRF, id_XPRF))
+        return id_XRF;
+    else if (cell_type.in(id_DDE, id_TDE, id_CDC, id_BGC, id_GBC))
+        return id_CDC;
+    else if (cell_type.in(id_XCDC))
+        return id_XCDC;
+    else if (cell_type.in(id_FIFO))
+        return id_FIFO;
+    else if (cell_type.in(id_XHFIFO, id_XWFIFO))
+        return id_XFIFO;
+    else if (cell_type.in(id_WFB, id_WFG))
+        return id_WFG;
+    else
+        return cell_type;
+}
+
+BelBucketId NgUltraImpl::getBelBucketForBel(BelId bel) const
+{
+    IdString bel_type = ctx->getBelType(bel);
+    if (bel_type.in(id_IOP, id_IOTP))
+        return ctx->idf("IOP/IOTP");
+    else if (bel_type.in(id_DFR, id_DDFR))
+        return ctx->idf("DFR/DDFR");
+    return bel_type;
+}
+
+bool NgUltraImpl::isValidBelForCellType(IdString cell_type, BelId bel) const
+{
+    IdString bel_type = ctx->getBelType(bel);
+    if (bel_type == id_IOTP)
+        return cell_type.in(id_IOP,id_IP,id_OP,id_IOTP,id_ITP,id_OTP);
+    else if (bel_type == id_IOP)
+        return cell_type.in(id_IOP,id_IP,id_OP);
+    else if (bel_type == id_DDFR)
+        return cell_type.in(id_BFR,id_DFR,id_DDFR);
+    else if (bel_type == id_DFR)
+        return cell_type.in(id_BFR,id_DFR);
+    else if (bel_type == id_RF)
+        return cell_type.in(id_RF,id_RFSP);
+    else if (bel_type == id_XRF)
+        return cell_type.in(id_XHRF,id_XWRF,id_XPRF);
+    else if (bel_type == id_CDC)
+        return cell_type.in(id_DDE, id_TDE, id_CDC, id_BGC, id_GBC);
+    else if (bel_type == id_XCDC)
+        return cell_type.in(id_XCDC);
+    else if (bel_type == id_FIFO)
+        return cell_type.in(id_FIFO);
+    else if (bel_type == id_XFIFO)
+        return cell_type.in(id_XHFIFO, id_XWFIFO);
+    else if (bel_type == id_WFG)
+        return cell_type.in(id_WFB,id_WFG);
+    else
+        return (bel_type == cell_type);
+}
+
+bool NgUltraImpl::getChildPlacement(const BaseClusterInfo *cluster, Loc root_loc,
+                                    std::vector<std::pair<CellInfo *, BelId>> &placement) const
+{
+    Loc prev = root_loc;
+    for (auto child : cluster->constr_children) {
+        Loc child_loc = if_using_basecluster<Loc>(child, [&](const BaseClusterInfo *child) {
```

Comment:
We don't need the whole `if_using_basecluster` conditional cast magic here, that's a `BaseArch` detail that allows `BaseArch` to not fail to compile if someone doesn't use `BaseClusterInfo` as it must support various kinds of arches.

Here for himbaechel we know that `ArchCellInfo` dervies from `BaseClusterInfo` so we can just access the relevant fields directly.

Reply:
This one is cleanedup, formatting will be covered by clangformat later,  made changes in way easier to review